sklearnex/decomposition/pca.py

```diff
@@ -15,4 +15,313 @@
+            elif sklearn_check_version("1.1"):
+                check_scalar(
+                    self.n_oversamples,
+                    "n_oversamples",
+                    min_val=1,
+                    target_type=numbers.Integral,
+                )
+
+            X = self._validate_data(
+                X,
+                dtype=[np.float64, np.float32],
+                ensure_2d=True,
+                copy=self.copy,
+            )
+
+            U, S, Vt = dispatch(
+                self,
+                "fit",
+                {
+                    "onedal": self.__class__._onedal_fit,
+                    "sklearn": sklearn_PCA._fit,
+                },
+                X,
+            )
+            return U, S, Vt
+
+        @run_with_n_jobs
+        def _onedal_fit(self, X, queue=None):
+            if issparse(X):
+                raise TypeError(
+                    "PCA does not support sparse input. See "
+                    "TruncatedSVD for a possible alternative."
+                )
+
+            self._validate_n_components(self.n_components, X.shape[0], X.shape[1])
+
+            onedal_params = {
+                "n_components": self.n_components,
+                "is_deterministic": True,
+                "method": "cov",
+                "whiten": self.whiten,
+            }
+            self._onedal_estimator = onedal_PCA(**onedal_params)
+            self._onedal_estimator.fit(X, queue=queue)
+            self._save_attributes()
+
+            U = None
+            S = self.singular_values_
+            Vt = self.components_
+
+            return U, S, Vt
+
+        def transform(self, X, y=None):
+            X = self._validate_data(
+                X,
+                dtype=[np.float64, np.float32],
+                reset=False,
+            )
+
+            return dispatch(
+                self,
+                "transform",
+                {
+                    "onedal": self.__class__._onedal_transform,
+                    "sklearn": sklearn_PCA.transform,
+                },
+                X,
+            )
+
+        @run_with_n_jobs
+        def _onedal_transform(self, X, queue=None):
+            check_is_fitted(self)
+
+            self._validate_n_features_in_after_fitting(X)
+            if sklearn_check_version("1.0"):
+                self._check_feature_names(X, reset=False)
+
+            return self._onedal_estimator.predict(X, queue=queue)
+
+        def fit_transform(self, X, y=None):
+            U, S, Vt = self._fit(X)
+            if U is None:
+                # oneDAL PCA was fit
+                X_transformed = self.transform(X, y)
+                return X_transformed
+            else:
+                # Scikit-learn PCA was fit
+                U = U[:, : self.n_components_]
+
+                if self.whiten:
+                    U *= sqrt(X.shape[0] - 1)
+                else:
+                    U *= S[: self.n_components_]
+
+                return U

```

Comment:
could you please add unit tests for the fit_transform?
I guess we will have problems with dpnp/dpctl interop here, since we are not using dispatch func or other usm_array support decorators

Reply:
for sklearnex dispatching is being used through the calls to _fit and transform. Direct computations are done only for sklearn since the fit_tranform and transform has different computations in sklearn. But agree with having more tests.